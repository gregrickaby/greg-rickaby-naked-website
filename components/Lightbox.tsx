'use client'

import {
  IconArrowNarrowLeft,
  IconArrowNarrowRight,
  IconX
} from '@tabler/icons-react'
import {useCallback, useEffect, useMemo, useRef, useState} from 'react'

export interface LightboxProps {
  images: string[]
  onClose: () => void
  startIndex: number
}

/**
 * Lightbox component.
 */
export default function Lightbox({images, onClose, startIndex}: LightboxProps) {
  // State to track the current index of the image being displayed.
  const [currentIndex, setCurrentIndex] = useState(startIndex)

  // Refs to track touch start and end positions.
  const touchStartX = useRef(0)

  // Update the current index when startIndex changes.
  useEffect(() => {
    setCurrentIndex(startIndex)

    // Set no-scroll on the body when the lightbox is open.
    document.body.style.overflow = 'hidden'
    return () => {
      document.body.style.overflow = 'auto'
    }
  }, [startIndex])

  // Go to the previous image.
  const goPrev = useCallback(() => {
    setCurrentIndex(
      (prevIndex) => (prevIndex + images.length - 1) % images.length
    )
  }, [images.length])

  // Go to the next image.
  const goNext = useCallback(() => {
    setCurrentIndex((prevIndex) => (prevIndex + 1) % images.length)
  }, [images.length])

  // Determine if there are images.
  const hasImages = useMemo(() => images && images.length > 0, [images])

  // Handle touch start.
  const handleTouchStart = (e: React.TouchEvent<HTMLDialogElement>) => {
    touchStartX.current = e.touches[0].clientX
  }

  // Handle touch end.
  const handleTouchEnd = (e: React.TouchEvent<HTMLDialogElement>) => {
    const touchEndX = e.changedTouches[0].clientX
    if (touchStartX.current - touchEndX > 50) {
      goNext()
    } else if (touchStartX.current - touchEndX < -50) {
      goPrev()
    }
  }

  // Handle keyboard events.
  useEffect(() => {
    const handleKeyDown = (e: KeyboardEvent) => {
      switch (e.key) {
        case 'ArrowLeft':
          goPrev()
          break
        case 'ArrowRight':
          goNext()
          break
        case 'Escape':
          onClose()
          break
      }
    }

    window.addEventListener('keydown', handleKeyDown)
    return () => {
      window.removeEventListener('keydown', handleKeyDown)
    }
  }, [goNext, goPrev, onClose])

  // No images? Do not render the lightbox.
  if (!hasImages) {
    return null
  }

  return (
    <dialog
      className="fixed inset-0 z-10 flex h-screen w-screen items-center justify-center bg-black bg-opacity-95 p-8"
      data-testid="lightbox"
      onTouchEnd={handleTouchEnd}
      onTouchStart={handleTouchStart}
    >
      {/* Close button */}
      <button
        aria-label="Close image"
        className="absolute right-4 top-4 z-50 rounded-full bg-black p-1 text-2xl text-white hover:bg-zinc-800"
        data-testid="close-button"
        onClick={onClose}
      >
        <IconX stroke={2} />
      </button>
      <div className="relative flex h-full w-full flex-col items-center justify-center gap-4 rounded-lg md:flex-row">
        {/* Previous image button */}
        <button
          aria-label="Previous image"
          className="absolute left-4 z-50 rounded-full bg-black p-1 text-white hover:bg-zinc-800"
          data-testid="prev-button"
          onClick={goPrev}
        >
          <IconArrowNarrowLeft stroke={2} />
        </button>
        {/* Image display */}
        <div className="max-w-5xl">
          <img
            alt={`${currentIndex + 1} of ${images.length}`}
            className="rounded border-4 border-white shadow-2xl md:max-h-[720px] md:max-w-[1280px]"
            data-testid="image"
            loading="eager"
            src={images[currentIndex]}
          />
        </div>
        {/* Next image button */}
        <button
          aria-label="Next image"
          className="absolute right-4 z-50 rounded-full bg-black p-1 text-white hover:bg-zinc-800"
          data-testid="next-button"
          onClick={goNext}
        >
          <IconArrowNarrowRight stroke={2} />
        </button>
      </div>
    </dialog>
  )
}
