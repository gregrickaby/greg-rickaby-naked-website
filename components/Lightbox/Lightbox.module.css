.dialog {
  @apply fixed inset-0 z-10 flex h-screen w-screen items-center justify-center bg-black bg-opacity-95 p-8;
}

.close {
  @apply absolute right-4 top-4 z-50 rounded-full bg-black p-1 text-2xl text-white hover:bg-zinc-800;
}

.container {
  @apply relative flex h-full w-full flex-col items-center justify-center gap-4 rounded-lg md:flex-row;
}

.previous {
  @apply absolute left-4 z-50 rounded-full bg-black p-1 text-white hover:bg-zinc-800;
}

.figure {
  @apply flex max-w-5xl flex-col items-center justify-center;
}

.image {
  @apply rounded border-4 border-white shadow-2xl md:max-h-[720px] md:max-w-[1280px];
}

.caption {
  @apply text-center text-sm text-white;
}

.next {
  @apply absolute right-4 z-50 rounded-full bg-black p-1 text-white hover:bg-zinc-800;
}
