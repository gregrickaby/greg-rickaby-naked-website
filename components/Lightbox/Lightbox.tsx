'use client'

import {Threads} from '@/lib/api'
import {
  IconArrowNarrowLeft,
  IconArrowNarrowRight,
  IconX
} from '@tabler/icons-react'
import {useCallback, useEffect, useMemo, useRef, useState} from 'react'
import styles from './Lightbox.module.css'

export interface LightboxProps {
  photos: Threads
  onClose: () => void
  startIndex: number
}

/**
 * Lightbox component.
 */
export function Lightbox({photos, onClose, startIndex}: LightboxProps) {
  // State to track the current index of the image being displayed.
  const [currentIndex, setCurrentIndex] = useState(startIndex)

  // Refs to track touch start and end positions.
  const touchStartX = useRef(0)

  // Update the current index when startIndex changes.
  useEffect(() => {
    setCurrentIndex(startIndex)

    // Set no-scroll on the body when the lightbox is open.
    document.body.style.overflow = 'hidden'
    return () => {
      document.body.style.overflow = 'auto'
    }
  }, [startIndex])

  // Go to the previous image.
  const goPrev = useCallback(() => {
    setCurrentIndex(
      (prevIndex) => (prevIndex + photos.data.length - 1) % photos.data.length
    )
  }, [photos.data.length])

  // Go to the next image.
  const goNext = useCallback(() => {
    setCurrentIndex((prevIndex) => (prevIndex + 1) % photos.data.length)
  }, [photos.data.length])

  // Determine if there are photos to display.
  const hasData = useMemo(
    () => photos.data && photos.data.length > 0,
    [photos.data]
  )

  // Handle touch start.
  const handleTouchStart = (e: React.TouchEvent<HTMLDialogElement>) => {
    touchStartX.current = e.touches[0].clientX
  }

  // Handle touch end.
  const handleTouchEnd = (e: React.TouchEvent<HTMLDialogElement>) => {
    const touchEndX = e.changedTouches[0].clientX
    if (touchStartX.current - touchEndX > 50) {
      goNext()
    } else if (touchStartX.current - touchEndX < -50) {
      goPrev()
    }
  }

  // Handle keyboard events.
  useEffect(() => {
    const handleKeyDown = (e: KeyboardEvent) => {
      switch (e.key) {
        case 'ArrowLeft':
          goPrev()
          break
        case 'ArrowRight':
          goNext()
          break
        case 'Escape':
          onClose()
          break
      }
    }

    window.addEventListener('keydown', handleKeyDown)
    return () => {
      window.removeEventListener('keydown', handleKeyDown)
    }
  }, [goNext, goPrev, onClose])

  // No photos.data? Do not render the lightbox.
  if (!hasData) {
    return null
  }

  return (
    <dialog
      className={styles.dialog}
      data-testid="lightbox"
      onTouchEnd={handleTouchEnd}
      onTouchStart={handleTouchStart}
    >
      <button
        aria-label="Close image"
        className={styles.close}
        data-testid="close-button"
        onClick={onClose}
      >
        <IconX stroke={2} />
      </button>

      <div className={styles.container}>
        <button
          aria-label="Previous image"
          className={styles.previous}
          data-testid="previous-button"
          onClick={goPrev}
        >
          <IconArrowNarrowLeft stroke={2} />
        </button>

        <figure className={styles.figure}>
          <a
            href={photos.data[currentIndex].permalink}
            title="View post on Threads.net"
          >
            <img
              alt={`${currentIndex + 1} of ${photos.data.length}`}
              className={styles.image}
              data-testid="image"
              loading="eager"
              src={photos.data[currentIndex].media_url}
            />
          </a>
          <figcaption className={styles.caption}>
            {photos.data[currentIndex].text}
          </figcaption>
        </figure>

        <button
          aria-label="Next image"
          className={styles.next}
          data-testid="next-button"
          onClick={goNext}
        >
          <IconArrowNarrowRight stroke={2} />
        </button>
      </div>
    </dialog>
  )
}
