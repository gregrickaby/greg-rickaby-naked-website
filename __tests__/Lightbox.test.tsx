import Lightbox from '@/components/Lightbox'
import '@testing-library/jest-dom'
import {act, fireEvent, render, waitFor} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {axe} from 'jest-axe'
import {useState} from 'react'
import {describe, expect, it} from 'vitest'

interface MockComponentProps {
  images: string[]
}

// Mock images.
const images = [
  'https://via.placeholder.com/600/92c952',
  'https://via.placeholder.com/600/771796',
  'https://via.placeholder.com/600/24f355'
]

// Mock component.
const MockComponent = ({images}: MockComponentProps) => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <>
      <button onClick={() => setIsOpen(!isOpen)}>Open</button>
      {isOpen && (
        <Lightbox
          images={images}
          onClose={() => setIsOpen(false)}
          startIndex={0}
        />
      )}
    </>
  )
}

describe('Lightbox', () => {
  afterEach(() => {
    document.body.style.overflow = 'auto'
  })

  it('should toggle the lightbox using state', async () => {
    const {getByRole, queryByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Verify the lightbox is rendered.
    await waitFor(() => {
      expect(queryByTestId('lightbox')).toBeInTheDocument()
    })

    // Close the lightbox.
    const closeButton = getByRole('button', {name: /open/i})
    await userEvent.click(closeButton)

    // Verify the lightbox is closed.
    await waitFor(() => {
      expect(queryByTestId('lightbox')).not.toBeInTheDocument()
    })
  })

  it('should close the lightbox when the close button is clicked', async () => {
    const {getByRole, queryByTestId, getByTestId} = render(
      <MockComponent images={images} />
    )

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Close the lightbox.
    const closeButton = getByTestId('close-button')
    await userEvent.click(closeButton)

    // Verify the lightbox is closed.
    await waitFor(() => {
      expect(queryByTestId('lightbox')).not.toBeInTheDocument()
    })
  })

  it('should navigate to the next image', async () => {
    const {getByRole, getByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Navigate to the next image.
    const nextButton = getByTestId('next-button')
    await userEvent.click(nextButton)

    // Verify the next image is displayed.
    await waitFor(() => {
      expect(getByTestId('image')).toHaveAttribute('src', images[1])
    })
  })

  it('should navigate to the next image using the next arrow key', async () => {
    const {getByRole, getByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Navigate to the next image.
    await act(async () => {
      window.dispatchEvent(new KeyboardEvent('keydown', {key: 'ArrowRight'}))
    })

    // Verify the next image is displayed.
    await waitFor(() => {
      expect(getByTestId('image')).toHaveAttribute('src', images[1])
    })
  })

  it('should navigate to the next image using a swipe gesture', async () => {
    const {getByRole, getByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Swipe to the next image.
    const dialog = getByTestId('lightbox')
    fireEvent.touchStart(dialog, {touches: [{clientX: 300}]})
    fireEvent.touchEnd(dialog, {changedTouches: [{clientX: 0}]})

    // Verify the next image is displayed.
    await waitFor(() => {
      expect(getByTestId('image')).toHaveAttribute('src', images[1])
    })
  })

  it('should navigate to the previous image', async () => {
    const {getByRole, getByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Navigate to the previous image.
    const prevButton = getByTestId('prev-button')
    await userEvent.click(prevButton)

    // Verify the previous image is displayed.
    await waitFor(() => {
      expect(getByTestId('image')).toHaveAttribute('src', images[2])
    })
  })

  it('should navigate to the previous image using a swipe gesture', async () => {
    const {getByRole, getByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Navigate to the previous image.
    const dialog = getByTestId('lightbox')
    fireEvent.touchStart(dialog, {touches: [{clientX: 0}]})
    fireEvent.touchEnd(dialog, {changedTouches: [{clientX: 300}]})

    // Verify the previous image is displayed.
    await waitFor(() => {
      expect(getByTestId('image')).toHaveAttribute('src', images[2])
    })
  })

  it('should navigate to the previous image using the previous arrow key', async () => {
    const {getByRole, getByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Navigate to the previous image.
    await act(async () => {
      window.dispatchEvent(new KeyboardEvent('keydown', {key: 'ArrowLeft'}))
    })

    // Verify the previous image is displayed.
    await waitFor(() => {
      expect(getByTestId('image')).toHaveAttribute('src', images[2])
    })
  })

  it('should close the lightbox when the escape key is pressed', async () => {
    const {getByRole, queryByTestId} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Close the lightbox.
    await act(async () => {
      window.dispatchEvent(new KeyboardEvent('keydown', {key: 'Escape'}))
    })

    // Verify the lightbox is closed.
    await waitFor(() => {
      expect(queryByTestId('lightbox')).not.toBeInTheDocument()
    })
  })

  it('should not render the lightbox if there are no images', async () => {
    const {getByRole, queryByTestId} = render(<MockComponent images={[]} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Verify the lightbox is not in the document.
    await waitFor(() => {
      expect(queryByTestId('lightbox')).not.toBeInTheDocument()
    })
  })

  it('should have no accessibility violations', async () => {
    const {container, getByRole} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    act(() => {
      userEvent.click(button)
    })

    // Verify the lightbox has no accessibility violations.
    await waitFor(async () => {
      const results = await axe(container)
      expect(results).toHaveNoViolations()
    })
  })

  it('should match the snapshot', async () => {
    const {container, getByRole} = render(<MockComponent images={images} />)

    // Open the lightbox.
    const button = getByRole('button', {name: /open/i})
    await userEvent.click(button)

    // Verify the rendered component matches the snapshot.
    await waitFor(() => {
      expect(container).toMatchSnapshot()
    })
  })
})
