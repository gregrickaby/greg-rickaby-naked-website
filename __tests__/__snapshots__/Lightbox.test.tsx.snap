// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lightbox > should match the snapshot 1`] = `
<div>
  <button>
    Open
  </button>
  <dialog
    class="fixed inset-0 z-10 flex h-screen w-screen items-center justify-center bg-black bg-opacity-95 p-8"
    data-testid="lightbox"
  >
    <button
      aria-label="Close image"
      class="absolute right-4 top-4 z-50 rounded-full bg-black p-1 text-2xl text-white hover:bg-zinc-800"
      data-testid="close-button"
    >
      <svg
        class="tabler-icon tabler-icon-x "
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 6l-12 12"
        />
        <path
          d="M6 6l12 12"
        />
      </svg>
    </button>
    <div
      class="relative flex h-full w-full flex-col items-center justify-center gap-4 rounded-lg md:flex-row"
    >
      <button
        aria-label="Previous image"
        class="absolute left-4 z-50 rounded-full bg-black p-1 text-white hover:bg-zinc-800"
        data-testid="prev-button"
      >
        <svg
          class="tabler-icon tabler-icon-arrow-narrow-left "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12l14 0"
          />
          <path
            d="M5 12l4 4"
          />
          <path
            d="M5 12l4 -4"
          />
        </svg>
      </button>
      <div
        class="max-w-5xl"
      >
        <img
          alt="1 of 3"
          class="rounded border-4 border-white shadow-2xl md:max-h-[720px] md:max-w-[1280px]"
          data-testid="image"
          loading="eager"
          src="https://via.placeholder.com/600/92c952"
        />
      </div>
      <button
        aria-label="Next image"
        class="absolute right-4 z-50 rounded-full bg-black p-1 text-white hover:bg-zinc-800"
        data-testid="next-button"
      >
        <svg
          class="tabler-icon tabler-icon-arrow-narrow-right "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12l14 0"
          />
          <path
            d="M15 16l4 -4"
          />
          <path
            d="M15 8l4 4"
          />
        </svg>
      </button>
    </div>
  </dialog>
</div>
`;
